File: R/Others/QRp.R
QRp<-function(X,Y,T,nu,solver){
# Quantile regression

LARGE<-1000000
n<-dim(Y)[1]
d<-dim(Y)[2]
r<-dim(X)[2]
m<-length(T)
U<-matrix(c(T[-1],1),m,1)
D <-diag(1,m);
for (i in 1:(m-1)) {D[i+1,i] <- (-1)}
mu <- D%*%U;
xbar <- t(nu)%*%X;

if ((n != dim(X)[1]) |( d != dim(U)[2] )) {stop("wrong dimensions")}
xbar <- t(nu)%*% X

############### VECTORIZATION OF THE MATRICES FOR LP ###############
c <- -kronecker(Y,mu)
A <- kronecker(t(X),sparseMatrix(1:m,1:m))
d <- matrix((matrix(1,m,1)-T)%*%xbar,nrow=m)
e <- matrix( matrix(1,m,1)%*% t(nu),nrow=m)

z_init <- matrix((matrix(1,m,1)-T)%*%t(nu),nrow=m)


############### LP SOLVING PHASE       ##############################
if (solver == 'gurobi')
{
model <- list()

model$A           <- A
model$obj         <- c
model$modelsense  <- "min"
model$rhs         <- d
model$ub          <- e
model$sense       <- '='
model$start		<- z_init
result		<- gurobi ( model, params=NULL ) 
if (result$status=="OPTIMAL") {Zvec <- result$x; L1vec <- t(result$pi); L2vec <- -t(Zvec>0) * t(result$rc) } else {stop("optimization problem with Gurobi")}

}
else
{
stop("Using MATLAB solver needs to be added")
}
#############################################

Z		<- matrix(Zvec,nrow=m)
L1		<- matrix(L1vec,nrow=m)
L2		<- matrix(L2vec,nrow=m)
beta		<- -diag(c(1/mu))%*% L1
P		<- -t(L2) %*% diag(c(1/mu))
val		<- t(mu) %*% Z %*% Y

 return(list(Z=Z,P=P,beta=beta, val=val))
}

File: Matlab/MKQR1D/QRp.m
function [ Z,P,beta, val ] = QRp( X,Y,T,nu, solver )
% Classical Quantile Regression
% computes standard classical regression 
% programmed by Alfred Galichon, with research assistance from Jinxin He
% reference: Roger Koenker; Gilbert Bassett, Jr. (1978). "Regression
% Quantiles". Econometrica, Vol. 46, No. 1. (Jan., 1978), pp. 33-50.

% --------------- construction of the constant matrices --------------
[n,d]=size(Y);
r=size(X,2);
m=length(T);
U=vertcat(T,1);
U(1,:)=[];
D = eye(m);
for i=1:m-1 D(i+1,i)=-1; end
mu=D*U;
xbar=nu'*X;
% -------------------------------------------------------------------
% --------------- vectorization of the matrices for LP --------------
c=-kron(Y,mu);
A=kron(X',sparse(1:m,1:m,1));
d=vec((ones(m,1)-T)*xbar);
e=vec(ones(m,1)*nu');
z_init=vec((ones(m,1)-T)*nu');
% -------------------------------------------------------------------
%-------------------  solves the LP ---------------------------------
if solver=='gurobi'

    model.A=sparse(A);
    model.obj=c;
    model.modelsense='min';
    model.rhs=d;
    model.ub=e;
    model.sense='=';
    model.start=z_init;
    result = gurobi(model);
    if strcmp(result.status, 'OPTIMAL')
        Zvec=result.x;
        L1vec=result.pi';
        L2vec=-(Zvec>0)' .* result.rc'; 
    else
        fprintf('Optimization returned status: %s\n', result.status);
    end
else if solver=='knitro'
        [Zvec, lambda, exitflag] = ktrlinklp (z_init, c, [], [], sparse(A), d, zeros(m,n), e );
        L1vec=-lambda.eqlin';
        L2vec=lambda.upper';
        
    else
        warning('off','MATLAB:rankDeficientMatrix');
        opts1= optimset('display','off');
        [Zvec,L1vec,L2vec]=lp_fnm(A, c', d, e, z_init);        
        % Solves a linear program by the interior point method:
        % min(c * z), s.t. A * z = d and 0 < z < e
    end
end
% -------------------------------------------------------------------
% ------ reshape sols into matrix form and computes results  --------

Z=reshape(Zvec,m,n);
L1=reshape(L1vec,m,r); % Lagrange mult associated to equality constraint
L2=reshape(L2vec,m,n); % Lagrange mult associated to inequality constraint

beta=-inv(diag(mu))*L1; % beta is the regression coefficient
P=-L2'*inv(diag(mu));   % P is the positive part of Y-X.beta
val=mu'*Z*Y;           % value of the program
end
